File: app/src/main/java/com/originb/inkwisenote2/modules/smartnotes/ui/TextNoteFragment.java
package com.originb.inkwisenote2.modules.smartnotes.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.originb.inkwisenote2.R;
import com.originb.inkwisenote2.common.Strings;
import com.originb.inkwisenote2.modules.backgroundjobs.BackgroundOps;
import com.originb.inkwisenote2.modules.backgroundjobs.Events;
import com.originb.inkwisenote2.modules.repositories.Repositories;
import com.originb.inkwisenote2.modules.repositories.SmartNotebook;
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity;
import com.originb.inkwisenote2.modules.smartnotes.data.NoteHolderData;
import com.originb.inkwisenote2.modules.textnote.data.TextNoteEntity;
import com.originb.inkwisenote2.modules.textnote.data.TextNotesDao;

import org.greenrobot.eventbus.EventBus;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

/**
 * Fragment for displaying and editing text notes
 */
public class TextNoteFragment extends NoteFragment {

    private EditText noteEditText;
    private ImageButton deleteBtn;
    private ImageButton debugButton;
    private TextNotesDao textNotesDao;
    private TextNoteEntity textNoteEntity;
    private File markdownFile;

    public TextNoteFragment(SmartNotebook smartNotebook, AtomicNoteEntity atomicNote) {
        super(smartNotebook, atomicNote);
        textNotesDao = Repositories.getInstance().getNotesDb().textNotesDao();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Inflate the layout for this fragment
        View itemView = inflater.inflate(R.layout.note_text_fragment, container, false);
        noteEditText = itemView.findViewById(R.id.note_edit_text);
        deleteBtn = itemView.findViewById(R.id.delete_note);
        debugButton = itemView.findViewById(R.id.debug_button);

        return itemView;
    }

    @Override
    public void onViewCreated(@NonNull View itemView, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(itemView, savedInstanceState);
        deleteBtn.setOnClickListener(view ->
                EventBus.getDefault()
                        .post(new Events.DeleteNoteCommand(smartNotebook,
                                atomicNote))
        );

        debugButton.setOnClickListener(v -> {
            showDebugDialog();
        });

        loadNote();
    }

    private void showDebugDialog() {
        if (getContext() != null) {
            NoteDebugDialog dialog = new NoteDebugDialog(getContext(), atomicNote, smartNotebook);
            dialog.show();
        }
    }

    protected void loadNote() {
        BackgroundOps.execute(() -> {
            // Check for markdown file first
            String markdownContent = loadMarkdownFile();
            // Get or create the text note entity for metadata
            textNoteEntity = textNotesDao.getTextNoteForNote(atomicNote.getNoteId());
            if (textNoteEntity == null) {
                textNoteEntity = new TextNoteEntity(atomicNote.getNoteId(), smartNotebook.smartBook.getBookId());
                textNotesDao.insertTextNote(textNoteEntity);
            }
            if (Strings.isNotEmpty(markdownContent) && !markdownContent.equals(textNoteEntity.getNoteText())) {
                textNoteEntity.setNoteText(markdownContent);
                textNotesDao.updateTextNote(textNoteEntity);
            } else if (Strings.isNullOrWhitespace(markdownContent)
                    && Strings.isNotEmpty(textNoteEntity.getNoteText())) {
                markdownContent = textNoteEntity.getNoteText();
            }

            return markdownContent;
        }, noteText -> {
            if (noteEditText != null) {
                noteEditText.setText(noteText);
            }
        });
    }

    /**
     * Load text from markdown file if it exists
     */
    private String loadMarkdownFile() {
        if (getContext() == null || Strings.isNullOrWhitespace(atomicNote.getFilepath())) return null;

        // Create markdown file path using notebook directory structure
        String markdownPath = getMarkdownFilePath();
        markdownFile = new File(markdownPath);

        if (!markdownFile.exists()) {
            return null;
        }

        StringBuilder content = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new FileReader(markdownFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                content.append(line).append("\n");
            }
            if (content.length() > 0) {
                content.deleteCharAt(content.length() - 1);
            }
            return content.toString();
        } catch (IOException e) {
            e.printStackTrace();
            if (getContext() != null) {
                Toast.makeText(getContext(), "Error reading note file", Toast.LENGTH_SHORT).show();
            }
            return null;
        }
    }

    @Override
    public NoteHolderData getNoteHolderData() {
        String text = noteEditText != null ? noteEditText.getText().toString().trim() : "";
        return NoteHolderData.textNoteData(text);
    }

    private String getMarkdownFilePath() {
        return atomicNote.getFilepath() + "/" + atomicNote.getFilename() + ".md";
    }
}

File: app/src/main/java/com/originb/inkwisenote2/modules/smartnotes/ui/TextNoteFragment.kt
package com.originb.inkwisenote2.modules.smartnotes.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageButton
import android.widget.Toast
import com.originb.inkwisenote2.R
import com.originb.inkwisenote2.common.isNotEmpty
import com.originb.inkwisenote2.common.isNullOrWhitespace
import com.originb.inkwisenote2.common.showDebugDialog
import com.originb.inkwisenote2.modules.backgroundjobs.BackgroundOps
import com.originb.inkwisenote2.modules.handwrittennotes.data.HandwrittenNoteRepository
import com.originb.inkwisenote2.modules.ocr.data.NoteOcrTextsDao
import com.originb.inkwisenote2.modules.repositories.SmartNotebook
import com.originb.inkwisenote2.modules.repositories.SmartNotebookRepository
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity
import com.originb.inkwisenote2.modules.smartnotes.data.NoteHolderData
import com.originb.inkwisenote2.modules.textnote.data.TextNoteEntity
import com.originb.inkwisenote2.modules.textnote.data.TextNotesDao
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException

/**
 * Fragment for displaying and editing text notes
 */
open class TextNoteFragment(
    smartNotebook: SmartNotebook?, atomicNote: AtomicNoteEntity?, private val textNotesDao: TextNotesDao,
// Additional dependencies for NoteDebugDialog
    private val handwrittenNoteRepository: HandwrittenNoteRepository?, private val noteOcrTextDao: NoteOcrTextsDao?,
    private val smartNotebookRepository: SmartNotebookRepository?
) : NoteFragment(smartNotebook, atomicNote) {
    private var noteEditText: EditText? = null
    private var deleteBtn: ImageButton? = null
    private var debugButton: ImageButton? = null
    private var textNoteEntity: TextNoteEntity? = null
    private var markdownFile: File? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        // Inflate the layout for this fragment
        val itemView = inflater.inflate(R.layout.note_text_fragment, container, false)
        noteEditText = itemView.findViewById<EditText?>(R.id.note_edit_text)
        deleteBtn = itemView.findViewById<ImageButton>(R.id.delete_note)
        debugButton = itemView.findViewById<ImageButton>(R.id.debug_button)

        return itemView
    }

    override fun onViewCreated(itemView: View, savedInstanceState: Bundle?) {
        super.onViewCreated(itemView, savedInstanceState)
        deleteBtn!!.setOnClickListener { confirmDeleteNote() }

        debugButton!!.setOnClickListener { _: View? ->
            showDebugDialog(
                context, atomicNote, smartNotebook,
                smartNotebookRepository, textNotesDao, noteOcrTextDao, handwrittenNoteRepository
            )
        }

        loadNote()
    }


    protected fun loadNote() {
        val note = atomicNote ?: return
        BackgroundOps.execute(
            {
                var markdownContent = loadMarkdownFile()
                textNoteEntity = textNotesDao.getTextNoteForNote(note.noteId)
                val entity = textNoteEntity!!
                if (isNotEmpty(markdownContent) && markdownContent != entity.noteText) {
                    entity.noteText = markdownContent
                    textNotesDao.updateTextNote(entity)
                } else if (isNullOrWhitespace(markdownContent) && isNotEmpty(entity.noteText)) {
                    markdownContent = entity.noteText ?: ""
                }
                markdownContent
            },
            { noteText ->
                noteEditText?.setText(noteText)
            }
        )
    }

    /**
     * Load text from markdown file if it exists
     */
    private fun loadMarkdownFile(): String {
        val note = atomicNote ?: return ""
        if (isNullOrWhitespace(note.filepath)) return ""

        // Create markdown file path using notebook directory structure
        val markdownPath = note.filepath + "/" + (note.filename ?: "") + ".md"
        markdownFile = File(markdownPath)

        if (!markdownFile!!.exists()) {
            return ""
        }

        val content = StringBuilder()
        try {
            BufferedReader(FileReader(markdownFile!!)).use { reader ->
                var line: String?
                while ((reader.readLine().also { line = it }) != null) {
                    content.append(line).append("\n")
                }
                if (content.isNotEmpty()) {
                    content.deleteCharAt(content.length - 1)
                }
                return content.toString()
            }
        } catch (e: IOException) {
            e.printStackTrace()
            context?.let { Toast.makeText(it, "Error reading note file", Toast.LENGTH_SHORT).show() }
            return ""
        }
    }

    override val noteHolderData: NoteHolderData
        get() {
            val text = noteEditText?.text?.toString()?.trim() ?: ""
            return NoteHolderData.textNoteData(text)
        }
}
